ÍNDICE:

1. Coincidencia de caracteres en expresiones regulares
2. Extrayendo datos usando expresiones regulares
3. Combinando búsqueda y extracción
4. Escapado de caracteres
5. Resumen
6. Sección adicional para usuarios de Unix/Linux
7. Depuración
8. Glosario
9. Ejercicios

In [3]:
import re

In [4]:
ruta = fr'Datos\\Python para todos\\mbox-short.txt'

manejador_archivos = open(ruta)

for linea in manejador_archivos:
    linea = linea.rstrip()
    if re.search('From:',linea):
        print(linea)

From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: gsilver@umich.edu
From: gsilver@umich.edu
From: zqian@umich.edu
From: gsilver@umich.edu
From: wagnermr@iupui.edu
From: zqian@umich.edu
From: antranig@caret.cam.ac.uk
From: gopal.ramasammycook@gmail.com
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: louis@media.berkeley.edu
From: ray@media.berkeley.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu


In [6]:
ruta = fr'Datos\\Python para todos\\mbox-short.txt'

manejador_archivos = open(ruta)

for linea in manejador_archivos:
    linea = linea.rstrip()
    if re.search('^From:',linea):
        print(linea)

From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: gsilver@umich.edu
From: gsilver@umich.edu
From: zqian@umich.edu
From: gsilver@umich.edu
From: wagnermr@iupui.edu
From: zqian@umich.edu
From: antranig@caret.cam.ac.uk
From: gopal.ramasammycook@gmail.com
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: louis@media.berkeley.edu
From: ray@media.berkeley.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu


# 1. Coincidencia de caracteres en expresiones regulares

In [5]:
ruta = fr'Datos\\Python para todos\\mbox-short.txt'

manejador_archivos = open(ruta)

for linea in manejador_archivos:
    linea = linea.rstrip()
    if re.search('^F..m',linea):
        print(linea)

From stephen.marquard@uct.ac.za Sat Jan  5 09:14:16 2008
From: stephen.marquard@uct.ac.za
From louis@media.berkeley.edu Fri Jan  4 18:10:48 2008
From: louis@media.berkeley.edu
From zqian@umich.edu Fri Jan  4 16:10:39 2008
From: zqian@umich.edu
From rjlowe@iupui.edu Fri Jan  4 15:46:24 2008
From: rjlowe@iupui.edu
From zqian@umich.edu Fri Jan  4 15:03:18 2008
From: zqian@umich.edu
From rjlowe@iupui.edu Fri Jan  4 14:50:18 2008
From: rjlowe@iupui.edu
From cwen@iupui.edu Fri Jan  4 11:37:30 2008
From: cwen@iupui.edu
From cwen@iupui.edu Fri Jan  4 11:35:08 2008
From: cwen@iupui.edu
From gsilver@umich.edu Fri Jan  4 11:12:37 2008
From: gsilver@umich.edu
From gsilver@umich.edu Fri Jan  4 11:11:52 2008
From: gsilver@umich.edu
From zqian@umich.edu Fri Jan  4 11:11:03 2008
From: zqian@umich.edu
From gsilver@umich.edu Fri Jan  4 11:10:22 2008
From: gsilver@umich.edu
From wagnermr@iupui.edu Fri Jan  4 10:38:42 2008
From: wagnermr@iupui.edu
From zqian@umich.edu Fri Jan  4 10:17:43 2008
From: zqian@

In [7]:
ruta = fr'Datos\\Python para todos\\mbox-short.txt'

manejador_archivos = open(ruta)

for linea in manejador_archivos:
    linea = linea.rstrip()
    if re.search('^From:.+@',linea):
        print(linea)

From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: gsilver@umich.edu
From: gsilver@umich.edu
From: zqian@umich.edu
From: gsilver@umich.edu
From: wagnermr@iupui.edu
From: zqian@umich.edu
From: antranig@caret.cam.ac.uk
From: gopal.ramasammycook@gmail.com
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: louis@media.berkeley.edu
From: ray@media.berkeley.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu


# 2. Extrayendo datos usando expresiones regulares

In [9]:
s = 'Una nota de csev@umich.edu a cwen@iupui.edu sobre una reunión @ 2PM'
lista = re.findall(r'\S+@\S+', s)
print(lista)

['csev@umich.edu', 'cwen@iupui.edu']


In [10]:
ruta = fr'Datos\\Python para todos\\mbox-short.txt'

manejador_archivos = open(ruta)

for linea in manejador_archivos:
    linea = linea.rstrip()
    x = re.findall(r'\S+@\S+',linea)
    if len(x) > 0:
        print(x)

['stephen.marquard@uct.ac.za']
['<postmaster@collab.sakaiproject.org>']
['<200801051412.m05ECIaH010327@nakamura.uits.iupui.edu>']
['<source@collab.sakaiproject.org>;']
['<source@collab.sakaiproject.org>;']
['<source@collab.sakaiproject.org>;']
['apache@localhost)']
['source@collab.sakaiproject.org;']
['stephen.marquard@uct.ac.za']
['source@collab.sakaiproject.org']
['stephen.marquard@uct.ac.za']
['stephen.marquard@uct.ac.za']
['louis@media.berkeley.edu']
['<postmaster@collab.sakaiproject.org>']
['<200801042308.m04N8v6O008125@nakamura.uits.iupui.edu>']
['<source@collab.sakaiproject.org>;']
['<source@collab.sakaiproject.org>;']
['<source@collab.sakaiproject.org>;']
['apache@localhost)']
['source@collab.sakaiproject.org;']
['louis@media.berkeley.edu']
['source@collab.sakaiproject.org']
['louis@media.berkeley.edu']
['louis@media.berkeley.edu']
['zqian@umich.edu']
['<postmaster@collab.sakaiproject.org>']
['<200801042109.m04L92hb007923@nakamura.uits.iupui.edu>']
['<source@collab.sakaiproject

In [11]:
ruta = fr'Datos\\Python para todos\\mbox-short.txt'

manejador_archivos = open(ruta)

for linea in manejador_archivos:
    linea = linea.rstrip()
    x = re.findall(r'[a-zA-Z0-9]\S+@\S+[a-zA-Z]', linea)
    if len(x) > 0:
        print(x)

['stephen.marquard@uct.ac.za']
['postmaster@collab.sakaiproject.org']
['200801051412.m05ECIaH010327@nakamura.uits.iupui.edu']
['source@collab.sakaiproject.org']
['source@collab.sakaiproject.org']
['source@collab.sakaiproject.org']
['apache@localhost']
['source@collab.sakaiproject.org']
['stephen.marquard@uct.ac.za']
['source@collab.sakaiproject.org']
['stephen.marquard@uct.ac.za']
['stephen.marquard@uct.ac.za']
['louis@media.berkeley.edu']
['postmaster@collab.sakaiproject.org']
['200801042308.m04N8v6O008125@nakamura.uits.iupui.edu']
['source@collab.sakaiproject.org']
['source@collab.sakaiproject.org']
['source@collab.sakaiproject.org']
['apache@localhost']
['source@collab.sakaiproject.org']
['louis@media.berkeley.edu']
['source@collab.sakaiproject.org']
['louis@media.berkeley.edu']
['louis@media.berkeley.edu']
['zqian@umich.edu']
['postmaster@collab.sakaiproject.org']
['200801042109.m04L92hb007923@nakamura.uits.iupui.edu']
['source@collab.sakaiproject.org']
['source@collab.sakaiproject

# 3. Combinando búsqueda y extracción

In [12]:
ruta = fr'Datos\\Python para todos\\mbox-short.txt'

manejador_archivos = open(ruta)

for linea in manejador_archivos:
    linea = linea.rstrip()
    if re.search(r'^X\S*: [0-9.]+',linea):
        print(linea)

X-DSPAM-Confidence: 0.8475
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.6178
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.6961
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.7565
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.7626
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.7556
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.7002
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.7615
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.7601
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.7605
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.6959
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.7606
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.7559
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.7605
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.6932
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.7558
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.6526
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.6948
X-DSPAM-Probability: 0.0000
X-DSPAM-Co

In [13]:
ruta = fr'Datos\\Python para todos\\mbox-short.txt'

manejador_archivos = open(ruta)

for linea in manejador_archivos:
    linea = linea.rstrip()
    x = re.findall(r'^X\S*: ([0-9.]+)',linea)
    if len(x) > 0:
        print(linea)

X-DSPAM-Confidence: 0.8475
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.6178
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.6961
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.7565
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.7626
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.7556
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.7002
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.7615
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.7601
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.7605
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.6959
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.7606
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.7559
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.7605
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.6932
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.7558
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.6526
X-DSPAM-Probability: 0.0000
X-DSPAM-Confidence: 0.6948
X-DSPAM-Probability: 0.0000
X-DSPAM-Co

In [14]:
ruta = fr'Datos\\Python para todos\\mbox-short.txt'

manejador_archivos = open(ruta)

for linea in manejador_archivos:
    linea = linea.rstrip()
    x = re.findall('^Details:.*rev=([0-9.]+)', linea)
    if len(x) > 0:
        print(x)

['39772']
['39771']
['39770']
['39769']
['39766']
['39765']
['39764']
['39763']
['39762']
['39761']
['39760']
['39759']
['39758']
['39757']
['39756']
['39755']
['39754']
['39753']
['39752']
['39751']
['39750']
['39749']
['39746']
['39745']
['39744']
['39743']
['39742']


In [15]:
ruta = fr'Datos\\Python para todos\\mbox-short.txt'

manejador_archivos = open(ruta)

for linea in manejador_archivos:
    linea = linea.rstrip()
    x = re.findall('^From .* ([0-9][0-9]):', linea)
    if len(x) > 0: 
        print(x)

['09']
['18']
['16']
['15']
['15']
['14']
['11']
['11']
['11']
['11']
['11']
['11']
['10']
['10']
['10']
['09']
['07']
['06']
['04']
['04']
['04']
['19']
['17']
['17']
['16']
['16']
['16']


# 4. Escapado de caracteres

In [16]:
x = 'Acabamos de recibir $10,00 para galletas.'
y = re.findall('\$[0-9.]+',x)
y

['$10']

# 5. Resumen

# 6. Sección adicional para usuarios de Unix/Linux

# 7. Depuración

In [ ]:
dir(re)

In [ ]:
help(re.search)

# 8. Glosario

Código frágil: código que funciona cuando los datos se encuentran en un formato específico pero tiene a romperse si éste no se cumple. Lo llamamos “código frágil” porque se rompe con facilidad.

Coincidencia ambiciosa: La idea de que los caracteres + y * en una expresión regular se expanden hacia afuera para coincidir con la mayor cadena posible.

Grep: un comando disponible en la mayoría de los sistemas Unix que busca en archivos de texto, buscando líneas que coincidan con expresiones regulares. El nombre del comando significa “Generalized Regular Expression Parser, o bien” Analizador Generalizado de Expresiones Regulares“.

Expresión regular: un lenguaje para encontrar cadenas más complejas en una búsqueda. Una expresión regular puede contener caracteres especiales que indiquen que una búsqueda solo coincida al comienzo o al final de una línea, junto con muchas otras capacidades similares.

Comodín: un carácter especial que coincide con cualquier carácter. En expresiones regulares, el carácter comodín es el punto.

# 9. Ejercicios

1. Escribe un programa simple que simule la operación del comando grep en Unix. Debe pedir al usuario que ingrese una expresión regular y cuente el número de líneas que coincidan con ésta:

$ python grep.py

Ingresa una expresión regular: ^Author

mbox.txt tiene 1798 líneas que coinciden con ^Author

$ python grep.py

Ingresa una expresión regular: ^X~

mbox.txt tiene 14368 líneas que coinciden con ^X~

$ python grep.py

Ingresa una expresión regular: java$

mbox.txt tiene 4175 líneas que coinciden con java$

In [18]:
# Grep.py
# Busque líneas que comiencen con From y tengan un signo de arroba

ruta = fr'Datos\\Python para todos\\mbox.txt'

manejador_archivos = open(ruta)

buscar = input('Introduzca una expresión regular: ')

contador = 0

for linea in manejador_archivos:
    linea = linea.rstrip()
    if re.search(buscar, linea): 
        contador += 1

print(f'mbox.txt tiene {contador} líneas que coincidieron con {buscar}' )

Introduzca una expresión regular: ^X
mbox.txt tiene 14368 líneas que coincidieron con ^X


2. Escribe un programa que busque líneas en la forma:

New Revision: 39772

Extrae el número de cada línea usando una expresión regular y el método findall(). Registra el promedio de esos números e imprímelo.

Ingresa nombre de archivo: mbox.txt

38444.0323119

Ingresa nombre de archivo: mbox-short.txt

39756.9259259

In [24]:
while True:
    nombre_archivo = input('Ingrese el nombre del archivo: ')
    ruta = fr'Datos\\Python para todos\\' + f'{nombre_archivo}'
    
    try:
        
        if nombre_archivo == 'fin':
            break
        
        manejador_archivos = open(ruta)
              
        buscar = 'New Revision: ([0-9]+)'
        contador = 0
        suma = 0

        for linea in manejador_archivos:
            linea = linea.strip()
            numero = re.findall(buscar, linea)
            if len(numero) > 0:
                suma += int(numero[0])
                contador += 1
        
        promedio = suma/contador
        print(promedio)
                
    except:
        print('El archivo no se encuentra, inténtelo de nuevo.')

Ingrese el nombre del archivo: mbox.txt
38549.79497206704
Ingrese el nombre del archivo: oipowipo
El archivo no se encuentra, inténtelo de nuevo.
Ingrese el nombre del archivo: mbox-short.txt
39756.92592592593
Ingrese el nombre del archivo: fin
